src/GHC/Util/HsExpr.hs

```diff
@@ -59,10 +59,15 @@ noSyntaxExpr' =
 dotApp' :: LHsExpr GhcPs -> LHsExpr GhcPs -> LHsExpr GhcPs
 dotApp' x y = noLoc $ OpApp noExt x (noLoc $ HsVar noExt (noLoc $ mkVarUnqual (fsLit "."))) y
 
+parSec' :: LHsExpr GhcPs -> LHsExpr GhcPs
+parSec' x@(LL _ SectionL{}) = noLoc $ HsPar noExt x
+parSec' x@(LL _ SectionR{}) = noLoc $ HsPar noExt x
+parSec' x = x
+
 dotApps' :: [LHsExpr GhcPs] -> LHsExpr GhcPs
 dotApps' [] = error "GHC.Util.HsExpr.dotApps', does not work on an empty list"
-dotApps' [x] = x
-dotApps' (x : xs) = dotApp' x (dotApps' xs)
+dotApps' [x] = parSec' x
+dotApps' (x : xs) = dotApp' (parSec' x) (dotApps' xs)
```

Comment:
The bracketing here is wrong.

Reply:
Don't disagree (it happens that it helps with a particular set of tests but is not principled). See my notes below.